internal/serialize/keyvalues.go

```diff
@@ -175,6 +182,16 @@ func ErrorToString(err error) (ret string) {
 	return
 }
 
+func writeMarshaler(b *bytes.Buffer, v logr.Marshaler) {
+	defer func() {
+		if err := recover(); err != nil {
+			// TODO: update tests. For now, behave as before.
+			writeStringValue(b, false, fmt.Sprintf("%+v", v))
+		}
+	}()
+	writeValue(b, v.MarshalLog())
```

Comment:
This could recurse if the result is itself a Marsahller - should this call a simplified version that doesn't do further expansion?

Reply:
I was wondering about that myself and came to the conclusion that recursion would be the right thing to do in this case. I don't know why MarshalLog should do this, but if it does, then the returned object is again something that wasn't meant to be logged as-is.

Reply:
It's not a huge deal, but I would err the other way - MarshalLog should never return something that needs further expansion.  I can't see why this would be required but I can see how it would be dangerous.

Reply:
I'm coming around to that point of view after thinking about what'll happen when MarshalLog (by mistake) returns itself... infinite recursion.

Looks like a good test case for https://github.com/kubernetes/klog/blob/main/test/output.go... will add that.